### PR TITLE
feat: route accroding to the graphql attributes

### PR DESCRIPTION
**Pull request by spacewander**
### Pre-submission checklist:

* [ ] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [ ] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

File: apisix/core/ctx.lua [hunk 1]
```diff
@@ -15,22 +15,96 @@
...
+    local ok, res = pcall(gq_parse, body)
+    if not ok then
+        return nil, "failed to parse graphql: " .. res .. " body: " .. body
+    end
+
+    if #res.definitions == 0 then
+        return nil, "empty graphql: " .. body
+    end
+
+    return res, nil
```
Comment (membphis) [inline apisix/core/ctx.lua:69]: `return  res` be fine

File: apisix/core/ctx.lua [hunk 2]
```diff
@@ -15,22 +15,96 @@
...
+end
+
+
+local function get_parsed_graphql(ctx)
+    if not ctx._graphql then
+        local res, err = parse_graphql(ctx)
+        if not res then
+            log.error(err)
+            ctx._graphql = {}
+
```
Comment (membphis) [inline apisix/core/ctx.lua:79]: return directly

File: apisix/core/ctx.lua [hunk 3]
```diff
@@ -84,6 +158,10 @@ do
                 key = re_gsub(key, "-", "_", "jo")
                 val = get_var(key, t._request)
 
+            elseif core_str.has_prefix(key, "graphql_") then
+                key = sub_str(key, 9)
```
Comment (membphis) [inline apisix/core/ctx.lua:162]: magic number, need comment

File: t/lib/test_admin.lua [hunk 4]
```diff
@@ -108,14 +108,15 @@ function _M.comp_tab(left_tab, right_tab)
     local err
     dir_names = {}
 
+    local _
     if type(left_tab) == "string" then
-        left_tab, err = json.decode(left_tab)
+        left_tab, _, err = json.decode(left_tab)
```
Comment (membphis) [inline t/lib/test_admin.lua:113]: where to get this API manual?

the new way seems wrong. I think the correct style should be `local res, err = json.decode(str)`

Comment (spacewander) [reply t/lib/test_admin.lua:113]: @membphis 
For dkjson, the second argument is `pos`, not `err`.

Comment (membphis) [reply t/lib/test_admin.lua:113]: got it, many thanks for your explain

**Comment by membphis** [comment]
@JanLi-air do you have time to look at this PR?

File: apisix/core/ctx.lua [hunk 5]
```diff
@@ -15,22 +15,96 @@
...
+    if #res.definitions == 0 then
+        return nil, "empty graphql: " .. body
+    end
+
+    return res
+end
+
+
+local function get_parsed_graphql(ctx)
+    if not ctx._graphql then
```
Comment (membphis) [inline apisix/core/ctx.lua:74]: code style

```lua
if ctx._graphql then
    return ctx._graphql
end

... ...
```

**Review by membphis — APPROVED** [review]
LGTM, except minor code style things

File: apisix/core/ctx.lua (hunk 5 above)
Comment (spacewander) [reply apisix/core/ctx.lua]: updated

**Review by moonming — APPROVED** [review]

**Pull request merged**